plugins/inputs/fritzbox/fritzbox.go

```diff
@@ -0,0 +1,402 @@
+//go:generate ../../../tools/readme_config_includer/generator
+package fritzbox
+
+import (
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/internal/choice"
+	"github.com/influxdata/telegraf/logger"
+	"github.com/influxdata/telegraf/plugins/common/tls"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/tdrn-org/go-tr064"
+	"github.com/tdrn-org/go-tr064/mesh"
+	"github.com/tdrn-org/go-tr064/services/igddesc/igdicfg"
+	"github.com/tdrn-org/go-tr064/services/tr64desc/deviceinfo"
+	"github.com/tdrn-org/go-tr064/services/tr64desc/hosts"
+	"github.com/tdrn-org/go-tr064/services/tr64desc/wancommonifconfig"
+	"github.com/tdrn-org/go-tr064/services/tr64desc/wandslifconfig"
+	"github.com/tdrn-org/go-tr064/services/tr64desc/wanpppconn"
+	"github.com/tdrn-org/go-tr064/services/tr64desc/wlanconfig"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+const pluginName = "fritzbox"
+
+const defaultTimeout = config.Duration(10 * time.Second)
+
+type Fritzbox struct {
+	URLs    []string        `toml:"urls"`
+	Collect []string        `toml:"collect"`
+	Timeout config.Duration `toml:"timeout"`
+	tls.ClientConfig
+	Log             telegraf.Logger `toml:"-"`
+	deviceClients   []*tr064.Client
+	serviceHandlers map[string]serviceHandlerFunc
+}
+
+func defaultFritzbox() *Fritzbox {
+	return &Fritzbox{
+		URLs:            make([]string, 0),
+		Collect:         []string{"device", "wan", "ppp", "dsl", "wlan"},
+		Timeout:         defaultTimeout,
+		deviceClients:   make([]*tr064.Client, 0),
+		serviceHandlers: make(map[string]serviceHandlerFunc),
+	}
+}
+
+func (*Fritzbox) SampleConfig() string {
+	return sampleConfig
+}
+
+func (plugin *Fritzbox) Init() error {
+	if plugin.Log == nil {
+		plugin.Log = logger.New("inputs", pluginName, "")
+	}
```

Comment:
Don't think this is needed, telegraf agent will only call Init when the logger is already set.

Reply:
This covers the case, when the plugin is called from Test code. Alternatevily the Logger may be initialized by the Test code as well.

Reply:
Yes, there's even a `testutil.Logger` for that case 😉